### PR TITLE
Add parser for flawfinder

### PR DESCRIPTION
**Pull request by davidCarlos**
I'm still working on this pull request (missing tests). soon i will remove the 'WIP' tag.

### CODE REVIEW

File: firehose/parsers/flawfinder.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+            while not prog.search(message_line) and message_line != "\n":
+                message += " " + message_line
+                line_cwe = get_cwe(message_line)
+                if line_cwe:
+                    cwe = line_cwe
+                message_line = loaded_file.readline()
+            weakness_messages.append(message)
+            # TODO: Flawfinder can returns more than one CWE,
+            # when this rappends, get_cwe cannot return a valid value
+            # Fix this.
```
Comment (davidmalcolm) [inline firehose/parsers/flawfinder.py]: Do we need to support more than one CWE per issue in the schema?  Currently <cwe> is optional within <issue>; should it be zeroOrMore?

Comment (davidCarlos) [inline firehose/parsers/flawfinder.py]: I was going to ask this. I don't know what the best pratice in this case. When flawfinder returns more than one CWE, what should i do?

Comment (davidmalcolm) [inline firehose/parsers/flawfinder.py]: Can you post some examples of what the output looks like?

My current opinion is that we should change the schema and model.py to support multiple CWEs per issue.

Comment (davidCarlos) [inline firehose/parsers/flawfinder.py]: ./docs/examples/rtsp.c:170:  [4] (buffer) sscanf:
  The scanf() family's %s operation, without a limit specification, permits
  buffer overflows (CWE-120, CWE-20). Specify a limit to %s, or use a
  different input function.
./docs/examples/synctime.c:155:  [4] (buffer) sscanf:
  The scanf() family's %s operation, without a limit specification, permits
  buffer overflows (CWE-120, CWE-20). Specify a limit to %s, or use a
  different input function.

In my opinion, Add multiple CWEs would generate a more realistic output, once one warning could be related to more than one CWE.

Comment (davidCarlos) [inline firehose/parsers/flawfinder.py]: I can send another PR doing this.

Comment (davidmalcolm) [inline firehose/parsers/flawfinder.py]: Thanks.  I'll update the schema accordingly.

Comment (davidCarlos) [inline firehose/parsers/flawfinder.py]: I will first make this change in models, then i finish the flawfinder parser.

Comment (davidCarlos) [inline firehose/parsers/flawfinder.py]: @davidmalcolm I have a doubt. Projects that already use firehose, and pass to Issue model a single cwe using a integer value. Should we maintain the compatibility? (Allows de user to pass a integer or a list of integers). Or we keep only a list of integers?

Comment (davidCarlos) [inline firehose/parsers/flawfinder.py]: This change impacts in the others parsers too.

**Comment by davidmalcolm** [comment]
As noted in #33 we don't have a good design yet for how to handle multiple CWEs in one issue.

For now, maybe have the parser capture the first CWE (if any), and note the limitation as a bug, marking it as a dependency of #33.

Thanks.

**Comment by davidCarlos** [comment]
Hey @davidmalcolm , i have updated this PR. I added some tests, and have included flawfinder in the documentation. Now the parser capture the first CWE, as we had discussed. I created the issue #35, to map the multiple cwes problem.

File: firehose/parsers/flawfinder.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                    cwe = extracted_cwe
+                message_line = loaded_file.readline()
+            weakness_messages.append(message)
+            weakness_cwes.append(cwe)
+            line = message_line
+        else:
+            line = loaded_file.readline()
+
+    counter = 0
+    for weakness in weakness_paths:
```
Comment (davidmalcolm) [inline firehose/parsers/flawfinder.py]: Rather than manually incrementing `counter`, could this code use Python's `enumerate` builtin?  (and maybe use `idx` rather then `counter`?)

File: firehose/parsers/flawfinder.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+        cwe = weakness_cwes[counter]
+        if cwe == '':
+            cwe = None
+        else:
+            cwe = int(weakness_cwes[counter])
+
+        issue = Issue(cwe, None, location,
+                      Message(text=weakness_messages[counter]), notes=None,
+                      trace=None, severity=None, customfields=None)
```
Comment (davidmalcolm) [inline firehose/parsers/flawfinder.py]: Re `severity`: Flawfinder's docs say:
> The risk level is shown inside square brackets and varies from 0, very little risk, to 5, great risk.

It would be good to capture that number as `Issue.severity` (as a string, without the square brackets):
http://firehose.readthedocs.io/en/latest/data-model.html#firehose.model.severity

File: firehose/parsers/flawfinder.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+                            function=None,
+                            point=Point(int(weakness_lines[counter]), 0))
+
+        cwe = weakness_cwes[counter]
+        if cwe == '':
+            cwe = None
+        else:
+            cwe = int(weakness_cwes[counter])
+
+        issue = Issue(cwe, None, location,
```
Comment (davidmalcolm) [inline firehose/parsers/flawfinder.py]: The `None` here is the `Issue.testid`:
http://firehose.readthedocs.io/en/latest/data-model.html#firehose.model.testid

Perhaps the `(shell) system:` stuff should be captured here.  If I'm reading the Flawfinder docs correctly:

> Each entry has a filename, a colon, a line number, a
>risk level in brackets (where 5 is the most risky), a category, the name of the function, and a description of

....then Flawfinder's "category" here is "shell" and the name of the function is "system".   I'm not yet sure how that should map to a firehose Issue.testid.

File: firehose/parsers/flawfinder.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+def parse_file(loaded_file):
+    """ Parser loaded flawfinder output
+
+    :loaded_file: flawfinder report, loaded in memory
+    :returns: Firehose Analysis object, representing the final XML.
+    """
+
+    generator = Generator(name='flawfinder',
+                          version=flawfinder())
```
Comment (davidCarlos) [inline firehose/parsers/flawfinder.py]: Fixed. Now i get the version from the Flawfinder output.

File: firehose/parsers/flawfinder.py (hunk 2 above)
Comment (davidCarlos) [inline firehose/parsers/flawfinder.py]: Fixed

File: firehose/parsers/flawfinder.py (hunk 4 above)
Comment (davidCarlos) [inline firehose/parsers/flawfinder.py]: @davidmalcolm how should we handle this? Apprently flawfinder does not have a testid value.

File: firehose/parsers/flawfinder.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        analysis.results.append(issue)
+        counter += 1
+
+    return analysis
+
+
+def get_cwe(line):
+    """Extract CWE from flawfinder alarms
+    It's possible that a alarm not have a correlated CWE
+
```
Comment (davidCarlos) [inline firehose/parsers/flawfinder.py]: Fixed

File: firehose/parsers/flawfinder.py (hunk 4 above)
Comment (davidmalcolm) [inline firehose/parsers/flawfinder.py]: Can we use flawfinder's  "category" here for the FIrehose "testid"?  Looking at the sample output, that would give values like "shell", "format", "buffer", "race", which seems sane.

**Comment by davidCarlos** [comment]
Hey @davidmalcolm , i updated this PR with the last revision.

File: firehose/parsers/flawfinder.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+    analysis = parse_file(report)
+    sys.stdout.write(str(analysis.to_xml()))
+    sys.stdout.write('\n')
+
+def load_file(arg_file):
+    """ Loads the Flawfinder report in memory
+
+    :arg1: file passed by cli
+    :returns: file loaded in memory
+    """
```
Comment (davidmalcolm) [inline firehose/parsers/flawfinder.py]: This doesn't actually load a file into memory; it merely returns a ``file`` object representing a file opened for reading.  The code would be clearer if you got rid of this function and replace users with just the ``open`` builtin.

File: firehose/parsers/flawfinder.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+        return open(arg_file, 'r')
+    except Exception as e:
+        raise e
+
+
+def parse_file(loaded_file):
+    """ Parser loaded flawfinder output
+
+    :loaded_file: flawfinder report, loaded in memory
+    :returns: Firehose Analysis object, representing the final XML.
```
Comment (davidmalcolm) [inline firehose/parsers/flawfinder.py]: "loaded_file" is actually a file-like object, not an in-memory string.  Please rename to "infile", or similar ("input" is a builtint IIRC)

File: firehose/parsers/flawfinder.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+            (prog.search(message_line) and prog2.search(message_line)):
+                message += " " + message_line
+                extracted_cwe = get_cwe(message_line)
+                if extracted_cwe != '':
+                    data["cwe"] = extracted_cwe
+                message_line = loaded_file.readline()
+            data["message"] = message;
+            line = message_line
+            flawfinder_data.append(data)
+        else:
```
Comment (davidmalcolm) [inline firehose/parsers/flawfinder.py]: I'm finding it hard to follow this two-stage parsing.  If I'm reading things right, the code first reads the lines, building up a list of dicts (``flawfinder_data``).  There's then a second loop over that list, creating ``Issue`` objects and adding them to ``analysis.results``.

Wouldn't it be much simpler to have one loop, and append ``Issue`` objects directly to ``analysis.results``, without the ``data`` dict i.e. work directly with an ``Issue`` object instead.

Hope that makes sense; I may have misread this code.

File: firehose/parsers/flawfinder.py [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+            the numeric part of the cwe
+            CWE-120, cwe_int_part will point to the character 1
+        """
+        cwe_first_int_char = cwe_init_str_pos + 4
+        try:
+            while 1:
+                int(line[cwe_first_int_char])
+                cwe_first_int_char += 1
+        except Exception:
+            cwe = line[(cwe_init_str_pos + 4):(cwe_first_int_char)]
```
Comment (davidmalcolm) [inline firehose/parsers/flawfinder.py]: Wouldn't this be much simpler with a regex  e.g.

     re.findall(line, "CWE-([0-9]+)")

or somesuch?  (which I think would give a list of tuples of strings).

Could then rename the function to ``find_cwes`` and have it return a list of ints, and do the dropping of multiples in the caller.

Can also write a unit test for the CWE parser that way.

File: firehose/parsers/flawfinder.py [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+    return cwe
+
+
+def flawfinder(first_line):
+    """Retrieve flawfinder version from report.
```
Comment (davidmalcolm) [inline firehose/parsers/flawfinder.py]: I find the current name to be unclear; please rename to ``parse_first_line`` (as that's what it does).

File: tests/parsers/test_flawfinder_parser.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+import os
+import unittest
+
+from firehose.parsers.flawfinder  import load_file,  parse_file
+from firehose.model import Analysis, Issue, Sut, Trace
+
+class TestParseXml(unittest.TestCase):
+    def parse_example(self, filename):
+        report = load_file(os.path.join(os.path.dirname(__file__),
```
Comment (davidmalcolm) [inline tests/parsers/test_flawfinder_parser.py]: Given that it's a file-like object, maybe turn into:

    path = os.path.join(...)
    with open(path) as f:
        return parse_file (f)

File: firehose/parsers/flawfinder.py (hunk 11 above)
Comment (davidCarlos) [inline firehose/parsers/flawfinder.py]: Maybe rename to `get_flawfinder_version` should be better, what do you think @davidmalcolm ?

**Comment by davidCarlos** [comment]
@davidmalcolm thanks for the great revision, i had updated the PR.

File: firehose/parsers/flawfinder.py [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+    File, Location, Generator, Metadata, Analysis, Issue
+
+
+def main():
+    """ Main entry to flawfinder parser """
+    try:
+        arg_file = sys.argv[1]
+        report = open(arg_file, 'r')
+    except IOError:
+        print("File note found")
```
Comment (davidmalcolm) [inline firehose/parsers/flawfinder.py]: Typo: "note" should be "not".
But an IOError is not necessarily "file not found"; it could be a permissions error, or an attempt to open a directory rather than a file, etc.   Maybe just lose this clause?

File: firehose/parsers/flawfinder.py [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+                                function=None,
+                                point=Point(int(issue_line), 0))
+
+            message_line = infile.readline()
+            issue_message = ""
+            while message_line != "\n" and not \
+            (prog.search(message_line) and prog2.search(message_line)):
+                issue_message += " " + message_line
+                try:
+                    cwe = re.findall("CWE-([0-9]+)", message_line)[0]
```
Comment (davidmalcolm) [inline firehose/parsers/flawfinder.py]: I don't like using try/except for parsing; I prefer to use non-exception program flow (rationale: sometimes we end up catching a different exception than the one we thought we were, so it can conceal bugs).

Anything that matches the ``([0-9]+)`` is known to be convertible to ``int``, so how about:

    # list of ints
    cwes = []

before the "while message_line" loop, and:

    cwes.extend([int(m.group(0))
                 for m in re.findall("CWE-([0-9]+)", message_line)])

here, and then after the "message_line" loop:

    if cwes:
       first_cwe = cwes[0]
    else
       first_cwe = None

to avoid the two try/except clauses.

Please reinstate the comment about issue #35

File: firehose/parsers/flawfinder.py [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+    prog = re.compile(pattern)
+    prog2 = re.compile(pattern2)
+
+
+    while line:
+        testid = None
+        try:
+            testid = re.findall(r"\(([a-z]+)\)", line)[0]
+        except IndexError:
+            pass
```
Comment (davidmalcolm) [inline firehose/parsers/flawfinder.py]: If we just want the first match, just use ``re.search``, which returns ``None`` for the "not-found" case ( https://docs.python.org/3.4/library/re.html#re.search ).

**Comment by davidCarlos** [comment]
@davidmalcolm  I had updated the PR

File: firehose/parsers/flawfinder.py [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+    prog2 = re.compile(pattern2)
+
+
+    while line:
+        if line != "\n" and (prog.search(line) and prog2.search(line)):
+
+            testid = re.search("\(([a-z]+)\)", line).group(1)
+            issue_path = line.split(":")[0]
+            issue_line = line.split(":")[1]
+
```
Comment (davidmalcolm) [inline firehose/parsers/flawfinder.py]: Thanks for doing the earlier cleanups; the code is clearer now.

With that, the handling of ``prog``, ``prog2`` and the ``testid`` looks a bit unusual; presumably these always appear on the same line, in a consistent way.  Given that, I think it would be improved by using a single regular expression; something like:

    # We want to match a line like:
    #    ./docs/examples/asiohiper.cpp:78:  [4] (shell) system:
    filename_group = "(.+)"
    linenum_group = "([0-9]+)
    severity_group = "\[([0-9])\]"
    category_group = "\((.+)\)"
    funcname_group = "(\S+)"
    ws = '\s+'
    PAT= (filename_group + ':' + linenum_group + ':' + ws + severity_group + ws
          + category_group + ws + funcname_group)

It should then be possible to do:

    m = re.match(PAT, message_line)
    if m:
      # do stuff with m.groups()

File: firehose/parsers/flawfinder.py [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+            line = message_line
+
+            if cwes:
+                first_cwe = int(cwes[0])
+            else:
+                first_cwe = None
+
+            issue = Issue(first_cwe, testid, location,
+                          Message(text=issue_message), notes=None,
+                          trace=None, severity=None, customfields=None)
```
Comment (davidmalcolm) [inline firehose/parsers/flawfinder.py]: Please set the ``severity`` field to the number in square parens (keep it as a string, rather than calling ``int`` on it).

File: firehose/parsers/flawfinder.py [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+            location = Location(file=File(issue_path, None),
+                                function=None,
+                                point=Point(int(issue_line), 0))
+
+            message_line = infile.readline()
+            issue_message = ""
+            cwes = []
+            while message_line != "\n" and not \
+            (prog.search(message_line) and prog2.search(message_line)):
+                issue_message += " " + message_line
```
Comment (davidmalcolm) [inline firehose/parsers/flawfinder.py]: You could reuse the PAT from above here.

**Comment by davidCarlos** [comment]
@davidmalcolm I had updated the PR
